### PR TITLE
bpo-35926: Add support for OpenSSL 1.1.1b on Windows

### PR DESCRIPTION
**Pull request by paulmon**
After a little more testing I discovered that these changes don't quite pass all of the tests yet.  
I will be looking into these failures tomorrow

amd64 retail - 1 failure
amd64 debug - 0 failures
win32 retail - 1 failures
win32 debug - access violation halts test

test_parse_cert_CVE_2019_5010 only fails win32 debug (access violation)
works for amd64 debug/release and win32 release

test_load_default_certs_env_windows fails on win32 and amd64 retail.  skipped on debug

https://bugs.python.org/issue35926

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Unfortunately we couldn't find an account corresponding to your GitHub username on [bugs.python.org](https://bugs.python.org/) (b.p.o) to verify you have signed the CLA (this might be simply due to a missing "GitHub Name" entry in your b.p.o account settings). This is necessary for legal reasons before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.


You can [check yourself](https://check-python-cla.herokuapp.com/) to see if the CLA has been received.

Thanks again for your contribution, we look forward to reviewing it!

File: PCbuild/openssl.vcxproj
Comment (paulmon) [inline PCbuild/openssl.vcxproj]: appveyor.yml uses VC-WIN64A-masm for this build.  Should this be updated?

**Comment by paulmon** [comment]
The access violation was fixed by a clean build.

The other failure looks like the value for x509 was incremented in the certificate store compared to the keycert.pem file that is checked in.   I'm still trying to learn what this means.

176 != 177,
AssertionError: {'x509': 176, 'crl': 0, 'x509_ca': 137} != {'x509': 177, 'crl': 0, 'x509_ca': 137}

**Comment by paulmon** [comment]
test_load_default_certs_env_windows is failing in my test build because after calling _wputenv(…) from EnvironmentVarGaurd() in python, when openssl calls getenv the environment variable isn't set so the key that was added to the path with the environment variables isn't counted...  still trying to figure out why this happens.  Could python and openssl be linking to different CRTs even though they are both retail?

File: PCbuild/openssl.vcxproj
Comment (zooba) [reply PCbuild/openssl.vcxproj]: Do you mean appveyor in the OpenSSL repository? If so, yeah, we should match them.

Comment (paulmon) [reply PCbuild/openssl.vcxproj]: yes that's what I meant

Comment (paulmon) [reply PCbuild/openssl.vcxproj]: Changing to VC-WIN64A-masm breaks the cpython build with a message that's not helpful:
Error: The operation could not be completed. Unspecified error 

If I'm reading the configure files correctly then "VC-WIN64A" is the 64-bit equivalent of "VC-WIN32 no-asm" and it works with the cpython build.

File: Lib/test/test_ssl.py
Comment (zooba) [inline Lib/test/test_ssl.py]: @tiran This would have been needed due to the recent CVE fix, right? Why didn't we have to put it in already? Or did we not add tests for this case.

File: Lib/test/test_ssl.py
Comment (zooba) [inline Lib/test/test_ssl.py]: Remember to reset this default once all the tests are fixed.

File: PCbuild/openssl.vcxproj
Comment (zooba) [reply PCbuild/openssl.vcxproj]: This file shouldn't be being used in the CPython build - it's used to build OpenSSL separately and then we dynamically link to it (I might consider moving it to `Tools` to better reflect this...). Maybe I'll have to try it and see what happens.

File: Lib/test/test_ssl.py
Comment (paulmon) [reply Lib/test/test_ssl.py]: This is the other half of the 'peer did not return a certificate' error change above.
The 'tlsv13 alert certificate required' below was not being raised because of the error in server.
Looking at the other code in this test I don't know if I changed the test correctly but I think it illustrates what I'm seeing.

File: PCbuild/openssl.vcxproj
Comment (paulmon) [reply PCbuild/openssl.vcxproj]: It's not used in the regular build but I needed to build openssl with the patch the prepare_ssl.bat applies before I could test changing to openssl 1.1.1a.   I assumed that this script is used to build the binaries for cpython-bin-deps based on the comments in PCBuild/readme.txt

**Comment by paulmon** [comment]
I ended up creating a new clone, rebuilding openssl with the current version of prepare_ssl, and building cpython clean.  When I run the tests for x86, and amd64, debug and retail for each they all currently pass.

File: Lib/test/test_ssl.py
Comment (paulmon) [reply Lib/test/test_ssl.py]: I reworked this so that chatty=False.
I think the new commit makes more sense.

File: PCbuild/openssl.vcxproj
Comment (zooba) [reply PCbuild/openssl.vcxproj]: Actually, this file isn't used at all for the official builds I've been doing ([example](https://dev.azure.com/python/cpython/_build/results?buildId=35360)). And it doesn't quite match how I've been building either... I'll take a look and see if I can get the build done.

File: Lib/test/test_ssl.py
Comment (tiran) [reply Lib/test/test_ssl.py]: Which CVE are you referring to? The CRL DP NULL pointer CVE? That shouldn't be related here.

With TLS 1.3 ``getpeercert`` returns None when the client has not send a client auth cert yet. This should only occur when the server-side has enable post handshake auth.

File: Lib/test/test_ssl.py
Comment (tiran) [inline Lib/test/test_ssl.py]: I wonder why I haven't run into this issue on Linux.

**Review by tiran — COMMENTED** [review]
Thanks for working on this. For some unknown reason, Windows behaves slightly different than Linux. I wonder what's going on.

https://github.com/python/cpython/pull/10031 adds some debug helpers to analyse the protocol. I'll try to land the PR first, then you can use it to diagnose the handshake.

**Comment by zooba** [comment]
I queued up a build on pipelines [here](https://dev.azure.com/Python/cpython/_build/results?buildId=37670) with the built binaries. It looked good apart from one thing that I'm about to comment on.

**Comment by zooba** [comment]
Correction - not going to point it out on the file because apparently you haven't touched the file?

`PCbuild/openssl.props` needs updating. In particular, `x64` is no longer added to the filenames for the 64-bit build (though `-arm` and `-arm64` are added for those builds).

**Comment by paulmon** [comment]
The -x64 is dropped when you change the openssl build type from VC-WIN64A to VC-WIN64A-masm. The props file seems to be the source of my mysterious build failure when I was trying out the VC-WIN64A-masm build.

**Comment by zooba** [comment]
I'm not seeing why Bedevere thinks the NEWS file is incorrectly formatted - it looks fine to me? I renamed the PR to see if that helps, but apparently not (or maybe it needs a new commit to reevaluate?)

Maybe try deleting and recreating the NEWS entry? Also specify in the text that this only updates Windows - @ned-deily will add his own entry for macOS.

**Comment by ned-deily** [comment]
I also don't see anything wrong with the news entry and, more importantly, neither does blurb when I run it locally.  I don't know what's going on on the bot side.  If it persists, we may need help from @Mariatta  or @brettcannon.

**Comment by Mariatta** [comment]
Looks fine, I'll try adding the skip news and removing it, in case it retrigger the check.

**Comment by Mariatta** [comment]
The news check failed because it is expecting the news entry to be more than 45 characters, so please re-write the news entry.

**Comment by Mariatta** [comment]
Edit: the actual news entry has to be at least 30 characters.

**Comment by ned-deily** [comment]
@Mariatta Can you say what is expecting the news entry to be that long and why?  AFAIK, neither blurb nor the docs build imposed a requirement.

**Comment by Mariatta** [comment]
@ned-deily There was discussion here: https://github.com/python/bedevere/issues/127

**Comment by ned-deily** [comment]
I guess we now have a counter-example. Personally, I don’t see a need to enforce a limit in the bot; in the end the release manager will review and edit the news entries anyway prior to release. OTOH I suppose the item could be padded.

**Comment by Mariatta** [comment]
🤔 add "Patch by Paul Monson."? 😅

I'm fine to tweak this to "as long as file not empty" if that's what we want.

**Comment by ned-deily** [comment]
That’s one solution!

I think the fact that we’ve all spent so much time on tracking this down and discussing it says to me that having the arbitrary limit is more trouble than it is worth.

**Comment by paulmon** [comment]
I'm seeing more test failures after switching to VC-WIN64A-masm.
They seem to be more of the handshake doesn't work the same  in OpenSSL 1.1.1 type with TLS 1.3 errors.
I see these being suppressed in an number of places even before I make changes

**Comment by Mariatta** [comment]
I've deployed the update to bedevere so the status will pass as long as the news is not empty.
It also appears that @paulmon had made the news file with 36 characters :)

File: Misc/NEWS.d/next/Security/2019-02-08-13-58-03.bpo-35926.uIEgyE.rst
Comment (Mariatta) [inline Misc/NEWS.d/next/Security/2019-02-08-13-58-03.bpo-35926.uIEgyE.rst]:
```suggestion
Update to OpenSSL 1.1.1a for Windows.
```

File: Lib/test/test_ssl.py
Comment (paulmon) [reply Lib/test/test_ssl.py]: I turned chatty on temporarily while testing and hit this code path which isn't normally hit.  I don't think it's new behavior

File: Lib/test/test_ssl.py
Comment (paulmon) [reply Lib/test/test_ssl.py]: The error is coming from Winsock.  Could there be a difference in the way Winsock handles certs and the way linux sockets handle certs that is causing this difference?

**Comment by tiran** [comment]
Please update to OpenSSL 1.1.1b.

**Comment by tiran** [comment]
Also see PR #12094

Since we have to update 3.7 to 1.1.1 in about half a year, I no longer mind to backport your improvements after 3.7.4 is out. It's @ned-deily's call.

**Comment by paulmon** [comment]
@zooba can you update the OpenSSL sources and binaries to OpenSSL 1.1.1b for Windows?

**Comment by paulmon** [comment]
Is there a right way to test openssl?
I have been using this command line, but I don't remember where I found it.
.\python.bat  -Werror -bb -m test -u urlfetch -u network -v test_ssl

**Comment by paulmon** [comment]
The last commit won't pass CI until cpython-bin-deps is updated to openssl 1.1.1b

File: Lib/test/test_ssl.py
Comment (paulmon) [reply Lib/test/test_ssl.py]: Most of the tests that are failing are failing in the same way as Linux.  I can tell because there a comments and expected exceptions in multiple places that say something like "sometimes connections fail with TLS 1.3"

The tests are opening a socket doing a handshake and the disconnecting the client before the session tickets are finished sending which results in various disconnect errors.  A real client would send something before closing the session, which appears to allow the session tickets to get sent.  In my mind these are bad tests.  

There are also these issues in the openssl issues about the session ticket protocol: https://github.com/openssl/openssl/issues/7948, https://github.com/openssl/openssl/issues/7967, https://github.com/python-trio/trio/issues/819 which don't really seem resolved.

What is needed to move this PR forward?

**Comment by zooba** [comment]
@paulmon I just pushed openssl-bin-1.1.1b branch (and sources) to the repo, so you should be unblocked here now.

File: Misc/NEWS.d/next/Security/2019-02-08-13-58-03.bpo-35926.uIEgyE.rst
Comment (ned-deily) [inline Misc/NEWS.d/next/Security/2019-02-08-13-58-03.bpo-35926.uIEgyE.rst]: Looks like this news file has been superseded by the one for 1.1.1b and should be deleted.

**Review by ned-deily — DISMISSED** [review]

File: Misc/NEWS.d/next/Security/2019-02-08-13-58-03.bpo-35926.uIEgyE.rst
Comment (paulmon) [reply Misc/NEWS.d/next/Security/2019-02-08-13-58-03.bpo-35926.uIEgyE.rst]: Thanks! This should be fixed now

**Comment by paulmon** [comment]
The code has been updated to OpenSSL 1.1.1b and all of the tests pass on my machine with the changes here to catch new exceptions.  What are the next steps for getting this merged?

**Comment by zooba** [comment]
I think we need @tiran to take another look and sign off.

**Comment by paulmon** [comment]
@ned-deily the stale .rst file has been removed as requested.
@tiran can you sign off?  Or is more work needed?
Thanks, Paul

File: Lib/test/test_ssl.py
Comment (zooba) [inline Lib/test/test_ssl.py]: Writing to standard output will cause the test to fail on buildbots (and then pass on rerun where stdout is expected).

If these errors are "okay", maybe we can handle them in `test_wrap_socket`?

File: Lib/test/test_ssl.py
Comment (zooba) [inline Lib/test/test_ssl.py]: Again, should not write to stdout.

File: Lib/test/test_ssl.py
Comment (zooba) [inline Lib/test/test_ssl.py]: Again

File: Lib/test/test_ssl.py
Comment (zooba) [reply Lib/test/test_ssl.py]: There's also a `support.verbose` flag that you can use to check whether additional output is expected.

File: Lib/test/test_ssl.py
Comment (zooba) [reply Lib/test/test_ssl.py]: Looking at the first of those issues, it seems like we should fix our tests to be negotiating the close properly.

**Review by zooba — COMMENTED** [review]
Couple of error messages that should be removed or put behind the verbose flag (or deliberately fail/skip/retry the test?), and still need @tiran's opinion on dealing with the new close behaviour in TLS 1.3

File: Lib/test/test_ssl.py
Comment (paulmon) [reply Lib/test/test_ssl.py]: removed the try/except completely

File: Lib/test/test_ssl.py
Comment (paulmon) [reply Lib/test/test_ssl.py]: This was caused by an error in openssl.  Removed try/except.
Submitted patches for openssl
https://github.com/openssl/openssl/pull/8590
https://github.com/python/cpython-source-deps/pull/13

File: Lib/test/test_ssl.py
Comment (paulmon) [reply Lib/test/test_ssl.py]: removed the try/except including the sys.stdout.write

File: Lib/test/test_ssl.py
Comment (paulmon) [reply Lib/test/test_ssl.py]: this change has been removed

File: Lib/test/test_ssl.py
Comment (paulmon) [reply Lib/test/test_ssl.py]: @tiran Why the OSError case is different on Linux I'm not sure.  On windows the OSError I was trying to handle was SSLError exception (based on OSError).  There is a race condition between the server getting the PEER_DID_NOT_RETURN_A_CERTIFICATE error and the client receiving the "tvs13 alert ..." error on Windows.  One of them always arrives but the protocol doesn't look like it guarantees enough synchronicity to force one of them to be always be first.


@zooba  The first 3 tests failures were caused by a bug in openssl so I have removed the try/excepts.  
I have submitted patches for openssl 
https://github.com/openssl/openssl/pull/8590
https://github.com/python/cpython-source-deps/pull/13

**Comment by zooba** [comment]
Still hoping for @tiran to take a look at the SSL test changes, but since I'll be in the same room as him in a couple of days I can "negotiate" more strongly.

I think we want the OpenSSL update for sure though, so getting this done ought to be easy enough.

**Comment by zooba** [comment]
@paulmon There's a new OpenSSL build available. Will that remove the need for the changes to the tests (i.e. we're throwing the right exception now)?

**Comment by paulmon** [comment]
That patch gets rid of the exceptions that were removed in this commit: https://github.com/python/cpython/pull/11779/commits/016a4857b51401afba3b9a8ba5684a619794dfd8

File: Lib/test/test_ssl.py
Comment (paulmon) [inline Lib/test/test_ssl.py:2353]: It seems that sockets are blocking by default on Linux, and non-blocking on Windows.  This could cause this difference, maybe. When I step through the socket creation code the socket is created on this line on Windows: 

    fd = WSASocketW(family, type, proto,
                           NULL, 0,
                           WSA_FLAG_OVERLAPPED | WSA_FLAG_NO_HANDLE_INHERIT);

WSA_FLAG_OVERLAPPED specifies a non-blocking socket.

Comment (paulmon) [reply Lib/test/test_ssl.py:2353]: I think the remaining errors are caused when the server side fails first, but would fail on the client first if the sockets were blocking

**Comment by zooba** [comment]
Okay, we may have to hold this for a bit longer.

I'll see if I can get some people who know this area to look it over at PyCon.

**Comment by paulmon** [comment]
> I'll see if I can get some people who know this area to look it over at PyCon.

I think that would be really helpful.  Thanks!

**Comment by zooba** [comment]
I've run the test suite a few times without issues, and I'm doing a custom buildbot run right now. If that comes up clear, I see no reason to keep from merging this.

**Comment by zooba** [comment]
@tiran Any last comments or concerns?

File: .travis.yml
Comment (zooba) [inline .travis.yml]: Undo this change:

```suggestion
    - OPENSSL=1.1.0i
```

**Comment by zooba** [comment]
Just noticed one more change to the Travis CI configuration. Which seems to work okay, but isn't part of enabling OpenSSL 1.1.1b _on Windows_ so we'll let Christian do it when he's ready.

**Comment by paulmon** [comment]
@zooba Makes sense to me. Merged your suggested change.

**Pull request merged**
